### PR TITLE
Support more url formats for Youtube video block

### PR DESCRIPTION
**Pull request by jomunker**
depends on #1206

This PR adds a better evaluation of different Youtube URL formats to the site block. Also Youtube links are now validated in the admin and a validation message is shown.

### Validation Error Screenshot
image

### CODE REVIEW

**Comment by johnnyomair** [comment]
Please add the validation of the URL to the API as well: You could use the `@Matches(<regex>)` from class-validator for it: https://github.com/typestack/class-validator

Also, when validating in both API and Admin, I think we could remove the validation from the site

**Comment by jomunker** [comment]
> Please add the validation of the URL to the API as well: You could use the `@Matches(<regex>)` from class-validator for it: https://github.com/typestack/class-validator
> 
> Also, when validating in both API and Admin, I think we could remove the validation from the site

I need the regex in site nevertheless since I have to parse the id from the url. I could remove the validation in the cms but then the admin user has no direct feedback. We need to discuss what behaviour we want exactly.

File: demo/site/src/blocks/YouTubeVideoBlock.tsx [hunk 1]
```diff
@@ -13,28 +13,49 @@ const getHeightInPercentForAspectRatio = (aspectRatio: YouTubeVideoBlockData["as
...
+const EXPECTED_YT_ID_LENGTH = 11;
+
+const parseYoutubeUrl = (url: string) => {
+    // regex from https://stackoverflow.com/a/27728417
+    const regExp = /^.*(?:(?:youtu\.be\/|v\/|vi\/|u\/\w\/|embed\/|shorts\/)|(?:(?:watch)?\?v(?:i)?=|&v(?:i)?=))([^#&?]*).*/;
+    const match = url.match(regExp);
+    const youtubeId = match && match[1].length == EXPECTED_YT_ID_LENGTH ? match[1] : null;
+
+    if (youtubeId) return youtubeId;
+    else throw new Error("Invalid YouTube URL");
```
[the remark this answers is not in the document]
Comment (jomunker) [reply demo/site/src/blocks/YouTubeVideoBlock.tsx:25]: will do thanks :)

**Comment by johnnyomair** [comment]
@jomunker please update/rebase this PR (the base PR has been merged with squash commit)

File: demo/site/src/blocks/YouTubeVideoBlock.tsx [hunk 2]
```diff
@@ -13,19 +13,23 @@ const getHeightInPercentForAspectRatio = (aspectRatio: YouTubeVideoBlockData["as
...
-        const searchParams = url.searchParams;
-        if (!searchParams.has("v")) {
-            throw new Error("URL has no ID (v) param");
-        }
-        youtubeIdentifier = searchParams.get("v") as string;
-    } catch (error) {
-        // no url, but ID was specified
-    }
+    const { success, identifier } = parseYoutubeUrl(youtubeIdentifier);
+    youtubeIdentifier = success && identifier ? identifier : "";
```
Comment (johnnyomair) [inline demo/site/src/blocks/YouTubeVideoBlock.tsx:32]: `success` isn't really useful here, you can return identifier from `parseYouTubeUrl` directly

**Review by johnnyomair — CHANGES_REQUESTED** [review]

File: demo/site/src/blocks/YouTubeVideoBlock.tsx [hunk 3]
```diff
@@ -13,19 +13,23 @@ const getHeightInPercentForAspectRatio = (aspectRatio: YouTubeVideoBlockData["as
...
 };
 
+const EXPECTED_YT_ID_LENGTH = 11;
+
+const parseYoutubeIdentifier = (value: string): string | undefined => {
+    // regex from https://stackoverflow.com/a/51870158
+    const regExp =
+        /(https?:\/\/)?(((m|www)\.)?(youtube(-nocookie)?|youtube.googleapis)\.com.*(v\/|v=|vi=|vi\/|e\/|embed\/|user\/.*\/u\/\d+\/)|youtu\.be\/)([_0-9a-zA-Z-]+)/;
+    const match = value.match(regExp);
+    const youtubeId = value.length === EXPECTED_YT_ID_LENGTH ? value : match && match[8].length == EXPECTED_YT_ID_LENGTH ? match[8] : null;
```
Comment (johnnyomair) [inline demo/site/src/blocks/YouTubeVideoBlock.tsx:23]: From the provided StackOverflow answer: "ID will always be in match group 8". Why do you use match group 9?

File: packages/admin/blocks-admin/src/blocks/YouTubeVideoBlock.tsx [hunk 4]
```diff
@@ -26,6 +42,8 @@ export const YouTubeVideoBlock: BlockInterface<YouTubeVideoBlockData, State, You
         return { ...state, autoplay: false, adminMeta: { route: previewCtx.parentUrl } };
     },
 
+    isValid: (state) => isValidYouTubeIdentifier(state.youtubeIdentifier),
```
Comment (johnnyomair) [inline packages/admin/blocks-admin/src/blocks/YouTubeVideoBlock.tsx:45]: Can't the identifier be `optional` here?

File: demo/site/src/blocks/YouTubeVideoBlock.tsx [hunk 5]
```diff
@@ -13,19 +13,23 @@ const getHeightInPercentForAspectRatio = (aspectRatio: YouTubeVideoBlockData["as
...
-        const searchParams = url.searchParams;
-        if (!searchParams.has("v")) {
-            throw new Error("URL has no ID (v) param");
-        }
-        youtubeIdentifier = searchParams.get("v") as string;
-    } catch (error) {
-        // no url, but ID was specified
-    }
+    const identifier = parseYoutubeIdentifier(youtubeIdentifier);
+    youtubeIdentifier = identifier ?? "";
```
Comment (johnnyomair) [inline demo/site/src/blocks/YouTubeVideoBlock.tsx:32]:
```suggestion
    youtubeIdentifier = parseYoutubeIdentifier(youtubeIdentifier) ?? "";
```

Should we really provide "" as default value here?

File: demo/site/src/blocks/YouTubeVideoBlock.tsx (hunk 3 above)
Comment (jomunker) [reply demo/site/src/blocks/YouTubeVideoBlock.tsx:23]: Maybe they confused match count and index or something. As you can see the match at index 8 is in fact the id.

image

File: packages/admin/blocks-admin/src/blocks/YouTubeVideoBlock.tsx (hunk 4 above)
Comment (jomunker) [reply packages/admin/blocks-admin/src/blocks/YouTubeVideoBlock.tsx:45]: No cause then the block would not make any sense I guess. Also there is a default value.

File: demo/site/src/blocks/YouTubeVideoBlock.tsx (hunk 5 above)
Comment (jomunker) [reply demo/site/src/blocks/YouTubeVideoBlock.tsx]: We need it for the url but i refactored it so it doesn't seem like a default value

**Review by johnnyomair — APPROVED** [review]

File: packages/admin/blocks-admin/src/blocks/YouTubeVideoBlock.tsx (hunk 4 above)
Comment (johnnyomair) [reply packages/admin/blocks-admin/src/blocks/YouTubeVideoBlock.tsx:45]: When you add an identifier and then clear the input field, the state of the identifier will be `undefined`. Furthermore, new blocks should be savable by default (see [docs](https://docs.comet-dxp.com/docs/blocks/block-best-practices#allow-saving-a-block-with-its-default-values)), and I don't think this works.

Comment (jomunker) [reply packages/admin/blocks-admin/src/blocks/YouTubeVideoBlock.tsx:46]: You're right. I fixed it.

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**